src/resolve/resolved-dns-stub.c

```diff
@@ -528,6 +603,73 @@ static int manager_dns_stub_tcp_fd(Manager *m) {
         return m->dns_stub_tcp_fd = TAKE_FD(fd);
 }
 
+static int manager_dns_stub_tcp_fd_extra(Manager *m, DNSStubListenerExtra *l) {
+        _cleanup_free_ char *pretty = NULL;
+        _cleanup_close_ int fd = -1;
+        int r;
+
+        if (l->fd >= 0)
+                return 0;
+
+        fd = socket(socket_address_family(&l->address), SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (fd < 0) {
+                r = -errno;
+                goto fail;
+        }
+
+        r = set_dns_stub_common_socket_options(fd);
+        if (r < 0)
+                goto fail;
+
+        r = setsockopt_int(fd, IPPROTO_IP, IP_TTL, true);
+        if (r < 0)
+                goto fail;
+
+        r = setsockopt_int(fd, IPPROTO_IP, IP_FREEBIND, true);
+        if (r < 0)
+                goto fail;
+
+        if (bind(fd, &l->address.sockaddr.sa, l->address.size) < 0) {
+                r = -errno;
+                goto fail;
+        }
+
+        if (listen(fd, SOMAXCONN) < 0) {
+                r = -errno;
+                goto fail;
+        }
+
+        r = sd_event_add_io(m->event, &l->dns_stub_extra_event_source, fd, EPOLLIN, on_dns_stub_packet, m);
+        if (r < 0)
+                goto fail;
+
+        (void) sd_event_source_set_description(l->dns_stub_extra_event_source, "dns-stub-tcp-extra");
+
+        l->fd = TAKE_FD(fd);
+
+        if (DEBUG_LOGGING) {
+                (void) sockaddr_pretty(&l->address.sockaddr.sa, FAMILY_ADDRESS_SIZE(l->address.sockaddr.sa.sa_family), true, true, &pretty);
+                log_debug("Listening on TCP socket %s.", strnull(pretty));
+        }
+
+        return 0;
+
+ fail:
+       (void) sockaddr_pretty(&l->address.sockaddr.sa, FAMILY_ADDRESS_SIZE(l->address.sockaddr.sa.sa_family), true, true, &pretty);
+       if (r == -EADDRINUSE)
+               return log_warning_errno(r,
+                                        "Another process is already listening on TCP socket %s.\n"
+                                        "Turning off local DNS stub extra support.", strnull(pretty));
+       else if (r == -EPERM)
+               return log_warning_errno(r,
+                                        "Failed to listen on TCP socket %s: %m.\n"
+                                        "Turning off local DNS stub extra support.", strnull(pretty));
+
+       assert(r < 0);
+
+       return log_error_errno(r, "Failed to listen on TCP socket %s: %m", strnull(pretty));
+}
```

Comment:
manager_dns_stub_tcp_fd_extra() is almost the same as manager_dns_stub_udp_fd_extra(). Please merge them, and just conditionalize on  `l->mode` internally:
```C
static int manager_dns_stub_fd_extra(Manager *m, DNSStubListenerExtra *l) {
        _cleanup_free_ char *pretty = NULL;
        _cleanup_close_ int fd = -1;
        int r;

        if (l->fd >= 0)
                return 0;

        bool udp = l->mode == DNS_STUB_LISTENER_UDP;
        type =  udp ? SOCK_DGRAM : SOCK_STREAM;
        fd = socket(socket_address_family(&l->address), type|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
        if (fd < 0) {
                r = -errno;
                goto fail;
        }

        r = set_dns_stub_common_socket_options(fd);
        if (r < 0)
                goto fail;

        if (!udp) {
             r = setsockopt_int(fd, IPPROTO_IP, IP_TTL, true);
             if (r < 0)
                    goto fail;
        }

        r = setsockopt_int(fd, IPPROTO_IP, IP_FREEBIND, true);
        if (r < 0)
                goto fail;

        if (bind(fd, &l->address.sockaddr.sa, l->address.size) < 0) {
                r = -errno;
                goto fail;
        }

        if (listen(fd, SOMAXCONN) < 0) {
                r = -errno;
                goto fail;
        }

         ...
}
```

set_dns_stub_common_socket_options() can then be folden into the new function, since it'll have just one user.

Reply:
```
static int manager_dns_stub_fd_extra(Manager *m, DNSStubListenerExtra *l) {
        _cleanup_free_ char *pretty = NULL;
        _cleanup_close_ int fd = -1;
        int r;

        if (l->fd >= 0)
                return 0;

        bool udp = l->mode == DNS_STUB_LISTENER_UDP;
        type =  udp ? SOCK_DGRAM : SOCK_STREAM;
        fd = socket(socket_address_family(&l->address), type|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
        if (fd < 0) {
                r = -errno;
                goto fail;
        }

        r = set_dns_stub_common_socket_options(fd);
        if (r < 0)
                goto fail;

```

It's looks similar in the initial part but then everywhere we have to put condition if we look down . UDP we don't to `listen` and the logging etc all needs condition. Which is I think not needed. ```set_dns_stub_common_socket_options() ``` has 4 users now.  This way it's much cleaner and other places also we have two functions for UDP and TCP.  
``` 
(void) sd_event_source_set_description(l->dns_stub_extra_event_source, "dns-stub-udp-extra"); 
                                                                                                                                                                                      
	if (DEBUG_LOGGING) {
                (void) sockaddr_pretty(&l->address.sockaddr.sa, FAMILY_ADDRESS_SIZE(l->address.sockaddr.sa.sa_family), true, true, &pretty);
                log_debug("Listening on UDP socket %s.", strnull(pretty));
        }

```